### PR TITLE
Seed data

### PR DESCRIPTION
**Pull request by EvalVis**
Adding seed data to assist debugging code.
However, something is not working from commit c1a260bd82a56392d8c6fea92e5b663c857d1fe1 (when I removed placeholders from repository), so I ask for assistance. Feel free to help.
Note that this update is necessary for backend programmers because without working repositories, no data transfer is possible, so before programming backend I suggest assisting here first.
In the last commit I remove old migrations (however, new migrations fail to build along with the code). Also changed some code in service class (because now Where statement is possible, when the issue is fixed).
When the code builds, it should seed the data (without hashed passwords because we don't have a hash implementation yet).

### CODE REVIEW

**Comment by EvalVis** [comment]
Fixed the project so it builds (razor files were causing silent build failures and (or) solution file). I have spotted two problems. First: manual Id insertion is necessary (at least in seed data file, maybe injection does something magical in SeedData.cs file database context is taken manually), ef core does not manage it itself, leaving an object without Id causes exception complaining that non-nullable constraint Id is null.
Another problem: Shipment seeding is unsuccessful. Everything in SeedData.cs was seeding successfully, there are couriers, senders (still without hashed passwords), post machines. Trying to seed shipments crashes the program. Probably because shipments have multiple parents, we will need to resolve this issue soon.

**Comment by EvalVis** [comment]
Solved the issue where shipments could not be seeded.

**Comment by EvalVis** [comment]
Also, added a demo in /courier page where all shipments and data relating to them can be seen (it's will not be implemented in production environment it's just for showing that db and seed data works).

**Comment by EvalVis** [comment]
Only one issue there in SeedData.cs Id must be specified to prevent crash is left. @yjhn can you check that out? Also, I have not tried it, but it is possible what this behaviour will also persist when repository is injected causing id non-null constraint violation exception.

**Comment by EvalVis** [comment]
Confirmed. Even with injection repository create method without Id set manually causes exception. Maybe ulong is not allowed and long is accepted?

**Comment by yjhn** [comment]
Rebase on main, I have fixed automatic ID generation and removed Name from Courier and merged #7.

**Comment by yjhn** [comment]
Rebased.

**Review by yjhn — CHANGES_REQUESTED** [review]
Wouldn't it be simpler and better to use csv instead of xlsx for seed data? CSV is much easier to edit. Also usage story should be in plaintext format, as it doesn't need any of Word's features.

**Comment by yjhn** [comment]
Also please add drawio source file for DB diagram so that everyone can edit it.

File: Persistence/Database/SeedData.cs [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+        public static void PopulateIfEmpty(IApplicationBuilder app)
+        {
+            ExcelPackage.LicenseContext = LicenseContext.NonCommercial;
+            DatabaseContext context = app.ApplicationServices.CreateScope().ServiceProvider
+                .GetRequiredService<DatabaseContext>();
+            if (migrate && context.Database.GetPendingMigrations().Any())
+            {
+                context.Database.Migrate();
+            }
```
Comment (yjhn) [inline Persistence/Database/SeedData.cs]: Migrations are already done in Program.cs. Currently migrations are always done, but control can be added. Also, this is not a good place to do migrations.

**Comment by yjhn** [comment]
I suggest adding repositories in another PR, as this is already big enough.

**Comment by EvalVis** [comment]
> Also please add drawio source file for DB diagram so that everyone can edit it.

The diagram source file is the file you see as .png. Just upload it and edit.

**Comment by EvalVis** [comment]
About the .xlsx and .csv: as a Windows user, I find myself more easy to work with .xlsx (because of the Excel interface obviously). Also .csv separates with commas, so commas become unusable. If you find yourself more at easy working with .csv I can add a configurable option to choose from these files, but the .xlsx stay.

**Comment by EvalVis** [comment]
Usage story is also left in .docx format because it has some colored formatting. It shows the features that should be but are not yet implemented. These could be changed to bold, however.

File: Persistence/Database/SeedData.cs (hunk 1 above)
Comment (EvalVis) [reply Persistence/Database/SeedData.cs]: Haven't seen it. Approved.

**Comment by EvalVis** [comment]
@yjhn Commit [304212c] fails, because in the post machines addresses we have comma like: [Street address], Vilnius.
Also, it fails with descriptions. Have you tried using it? I also thought about .csv because I was too tired to setup EPPlus, but when I found that problem and decided to implement EPPlus. Now .xlsx is superior.

File: Persistence/Database/SeedData.cs [hunk 2]
```diff
@@ -45,40 +49,38 @@ public static void PopulateIfEmpty(IApplicationBuilder app)
...
                     ulong id = Convert.ToUInt64(
                         sheet.Cells[i, idColumnPosition].Value.ToString());
                     string username = sheet.Cells[i, usernameColumnPosition].Value.ToString();
                     string password = sheet.Cells[i, passwordColumnPosition].Value.ToString();
                     // TODO: password hash.
-                    Sender sender = new Sender
-                    {
-                        Id = id, Username = username, HashedPassword = new byte[7]
-                    };
+                    Sender sender = new() { Id = id, Username = username, HashedPassword = new byte[7] };
```
Comment (EvalVis) [inline Persistence/Database/SeedData.cs:55]: Well, I think that is not a good formating (if you consider writing single line if in multiple lines), but it's not a problem for me.

File: Persistence/Database/SeedData.cs [hunk 3]
```diff
@@ -45,40 +49,38 @@ public static void PopulateIfEmpty(IApplicationBuilder app)
...
-            using (var excelFile = new ExcelPackage(excelFileInfo))
+            FileInfo excelFileInfo = new($"{basePath}/{nameof(Courier)}s.xlsx");
+            using (ExcelPackage excelFile = new ExcelPackage(excelFileInfo))
             {
-                var sheet = excelFile.Workbook.Worksheets.First();
-                int idColumnPosition = EPPlusExtensions.GetColumnByName(sheet, "ID");
-                int usernameColumnPosition = EPPlusExtensions.GetColumnByName(sheet, "Username");
-                int passwordColumnPosition = EPPlusExtensions.GetColumnByName(sheet, "Password");
-                int nameColumnPosition = EPPlusExtensions.GetColumnByName(sheet, "Name");
+                ExcelWorksheet? sheet = excelFile.Workbook.Worksheets.First();
```
Comment (EvalVis) [inline Persistence/Database/SeedData.cs:68]: Not necessary "?" This is for debugging, if it fails, programmers just need to check it out.

File: Extensions/EPPlusExtensions.cs [hunk 4]
```diff
@@ -4,12 +4,14 @@ namespace shipping_service.Extensions
 {
     public static class EPPlusExtensions
     {
-
         public static int GetColumnByName(this ExcelWorksheet ws, string columnName)
         {
-            if (ws == null) throw new ArgumentNullException(nameof(ws));
+            if (ws == null)
```
Comment (EvalVis) [inline Extensions/EPPlusExtensions.cs:9]: Consider what autoformating has done in SeedData.cs and this. These are not consistent.

File: Persistence/Database/SeedData.cs [hunk 5]
```diff
@@ -150,26 +147,27 @@ private static void populateShipments(DatabaseContext context)
...
+                    string shipmentStatusString =
                         sheet.Cells[i, statusColumnPosition].Value.ToString();
-                    ShipmentStatus shipmentStatus = 
-                        (ShipmentStatus) Enum.Parse(typeof(ShipmentStatus), shipmentStatusString);
-                    Shipment shipment = 
-                        new Shipment
+                    ShipmentStatus shipmentStatus =
+                        (ShipmentStatus)Enum.Parse(typeof(ShipmentStatus), shipmentStatusString);
+                    Shipment shipment =
+                        new()
```
Comment (EvalVis) [inline Persistence/Database/SeedData.cs:144]: I like the old stile, but the new one also fits to me.

**Comment by yjhn** [comment]
> The diagram source file is the file you see as .png. Just upload it and edit.

Didn't know you could do that, but it works.
As for CSVs, they now use `;` as a separator.

**Review by yjhn — APPROVED** [review]

**Comment by yjhn** [comment]
We won't use CSVs as it would require rewriting the parser for CSV.

File: Services/ShipmentService.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+namespace shipping_service.Services
+{
+    public class ShipmentService : IShipmentService
+    {
+        private readonly IShipmentRepository _packageRepository;
+
+        public ShipmentService(IShipmentRepository packageRepository)
+        {
+            _packageRepository = packageRepository;
+        }
```
Comment (RaimondSob) [inline Services/ShipmentService.cs]: inconsistent naming

File: Repositories/CourierRepository.cs [hunk 7]
```diff
@@ -1,32 +1,34 @@
...
 
-        public async Task<Courier> UpdateAsync(ulong id, Courier courierIn)
+        public async Task UpdateAsync(Courier courier)
         {
-            return new Courier();
+            await context.SaveChangesAsync();
         }
 
-        public async Task<ulong> DeleteAsync(ulong id)
+        public void DeleteAsync(Courier courier)
```
Comment (RaimondSob) [inline Repositories/CourierRepository.cs]: This should be either be made async, or remove Async in the name for less confusion.

File: Repositories/PostMachineRepository.cs [hunk 8]
```diff
@@ -1,32 +1,34 @@
...
 
-        public async Task<PostMachine> UpdateAsync(ulong id, PostMachine postMachineIn)
+        public async Task UpdateAsync(PostMachine postMachine)
         {
-            return new PostMachine();
+            await context.SaveChangesAsync();
         }
 
-        public async Task<ulong> DeleteAsync(ulong id)
+        public void DeleteAsync(PostMachine postMachine)
```
Comment (RaimondSob) [inline Repositories/PostMachineRepository.cs]: This should be either be made async, or remove Async in the name for less confusion.

File: Repositories/SenderRepository.cs [hunk 9]
```diff
@@ -1,32 +1,34 @@
...
 
-        public async Task<Sender> UpdateAsync(ulong id, Sender senderIn)
+        public async Task UpdateAsync(Sender sender)
         {
-            return new Sender();
+            await context.SaveChangesAsync();
         }
 
-        public async Task<ulong> DeleteAsync(ulong id)
+        public void DeleteAsync(Sender sender)
```
Comment (RaimondSob) [inline Repositories/SenderRepository.cs]: This should be either be made async, or remove Async in the name for less confusion.

File: Repositories/ShipmentRepository.cs [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+            await context.AddAsync(shipment);
+            await context.SaveChangesAsync();
+        }
+
+        public async Task UpdateAsync(Shipment shipment)
+        {
+            await context.SaveChangesAsync();
+        }
+
+        public void DeleteAsync(Shipment shipment)
```
Comment (RaimondSob) [inline Repositories/ShipmentRepository.cs]: This should be either be made async, or remove Async in the name for less confusion.

**Review by pskramtai — APPROVED** [review]

**Comment by EvalVis** [comment]
@RaimondSob Fixed as you requested.

**Review by vleveris — APPROVED** [review]

**Review by RaimondSob — APPROVED** [review]

**Pull request merged**